src/components/utils/include/utils/timer.h

```diff
@@ -84,18 +89,137 @@ class Timer {
+   * @param timeout_milliseconds new timeout value
+   *
+   */
+  virtual void UpdateTimeOut(const uint32_t timeout_milliseconds);
+
+  /**
    * @brief GetTimeout
    * @return returns timeout
    */
   Milliseconds GetTimeout() const;
 
  private:
+  /**
+   * @brief Delegate release timer, will call callback function one time
+   */
+  class TimerDelegate : public threads::ThreadDelegate {
+    friend class Timer;
+
+   public:
+    /**
+     * @brief Default constructor
+     *
+     * @param timer_thread The Timer_thread pointer
+     */
+    explicit TimerDelegate(Timer* timer);
+
+    /**
+     * @brief Destructor
+     */
+    virtual ~TimerDelegate();
+
+    /**
+     * @brief Thread main function.
+     */
+    void threadMain() OVERRIDE;
+
+    /**
+     * @brief Called by thread::thread to free all allocated resources.
+     */
+    void exitThreadMain() OVERRIDE;
+
+    /**
+     * @brief Set new Timeout
+     * @param timeout_milliseconds New timeout to be set
+     */
+    virtual void SetTimeOut(const uint32_t timeout_milliseconds);
+
+    /**
+      * @brief Wait until timer will start
+      */
+    virtual void WaitUntilStart();
+
+   protected:
+    Timer* timer_;
+    uint32_t timeout_milliseconds_;
+    sync_primitives::Lock state_lock_;
+    sync_primitives::ConditionalVariable termination_condition_;
+    sync_primitives::ConditionalVariable starting_condition_;
+    volatile bool stop_flag_;
+    sync_primitives::Lock restart_flag_lock_;
+    volatile bool restart_flag_;
+    volatile bool is_started_flag_;
+
+    /**
+     * @brief Gets timeout with overflow check
+     * @return timeout
+     */
+    inline int32_t Get_timeout() const {
+      return std::min(
+          static_cast<uint32_t>(std::numeric_limits<int32_t>::max()),
+          timeout_milliseconds_);
+    }
+
+   private:
+    /**
+     * @brief Quits threadMain function after next loop.
+     */
+    virtual void ShouldBeStoped();
+
+    /**
+     * @brief Restarts non-loop timer after current iteration.
+     */
+    virtual void ShouldBeRestarted();
+
+    /**
+     * @brief Return flag IsGoingStop
+     */
+    bool IsGoingStop();
+
+    DISALLOW_COPY_AND_ASSIGN(TimerDelegate);
+  };
+
+  /**
+   * @brief Delegate release looper timer.
+   * Will call delegate every timeout function while stop()
+   * won't be called
+   */
+  class TimerLooperDelegate : public TimerDelegate {
+   public:
+    /**
+     * @brief Default constructor
+     *
+     * @param timer_thread The Timer_thread pointer
+     * @param timeout      Timeout to be set
+     */
+    explicit TimerLooperDelegate(Timer* timer);
+
+    /**
+     * @brief Thread main function.
+     */
+    void threadMain() OVERRIDE;
+
+   private:
+    DISALLOW_COPY_AND_ASSIGN(TimerLooperDelegate);
+  };
+
   const std::string name_;
   const TimerTask* task_;
-  bool repeatable_;
   uint32_t timeout_ms_;
   bool is_running_;
-  timer_t timer_;
+  threads::Thread* thread_;
+  TimerDelegate* delegate_;
```

Comment:
On this pointer I'll create dynamically object of sub-class and it is thread-timer.